pilot/pkg/proxy/envoy/v2/monitoring.go

```diff
@@ -0,0 +1,206 @@
+	edsSendErrPushes  = pushes.WithTags(tag.Upsert(typeTag, "eds_senderr"))
+	ldsPushes         = pushes.WithTags(tag.Upsert(typeTag, "lds"))
+	ldsSendErrPushes  = pushes.WithTags(tag.Upsert(typeTag, "lds_senderr"))
+	ldsBuildErrPushes = pushes.WithTags(tag.Upsert(typeTag, "lds_builderr"))
+	rdsPushes         = pushes.WithTags(tag.Upsert(typeTag, "rds"))
+	rdsSendErrPushes  = pushes.WithTags(tag.Upsert(typeTag, "rds_senderr"))
+	rdsBuildErrPushes = pushes.WithTags(tag.Upsert(typeTag, "rds_builderr"))
+
+	pushErrors, pushErrorsView = monitoring.NewInt64AndView(
+		"pilot_xds_push_errors",
+		"Number of errors (timeouts) pushing to sidecars.",
+		view.Sum(),
+		typeTag)
+
+	unrecoverableErrs = pushErrors.WithTags(tag.Upsert(typeTag, "unrecoverable"))
+	retryErrs         = pushErrors.WithTags(tag.Upsert(typeTag, "retry"))
+
+	// only supported dimension is millis, unfortunately. default to unitdimensionless.
+	proxiesConvergeDelay, proxiesConvergeDelayView = monitoring.NewFloat64AndView(
+		"pilot_proxy_convergence_time",
+		"Delay between config change and all proxies converging.",
+		view.Distribution(1, 3, 5, 10, 20, 30, 50, 100),
+	)
+
+	pushContextErrors, pushContextErrorsView = monitoring.NewInt64AndView(
+		"pilot_xds_push_context_errors",
+		"Number of errors (timeouts) initiating push context.",
+		view.Sum())
+
+	totalXDSInternalErrors, totalXDSInternalErrorsView = monitoring.NewInt64AndView(
+		"pilot_total_xds_internal_errors",
+		"Total number of internal XDS errors in pilot.",
+		view.Sum())
+
+	inboundUpdates, inboundUpdatesView = monitoring.NewInt64AndView(
+		"pilot_inbound_updates",
+		"Total number of updates received by pilot.",
+		view.Sum(),
+		typeTag)
+
+	inboundConfigUpdates   = inboundUpdates.WithTags(tag.Upsert(typeTag, "config"))
+	inboundEDSUpdates      = inboundUpdates.WithTags(tag.Upsert(typeTag, "eds"))
+	inboundServiceUpdates  = inboundUpdates.WithTags(tag.Upsert(typeTag, "svc"))
+	inboundWorkloadUpdates = inboundUpdates.WithTags(tag.Upsert(typeTag, "workload"))
+)
+
+func incrementXDSRejects(stat *monitoring.Int64, node, errCode string) {
+	stat.WithTags(tag.Upsert(nodeTag, node), tag.Upsert(errTag, errCode)).Increment()
+	totalXDSRejects.Increment()
+}
+
+func init() {
+
+	if errTagErr != nil {
+		panic(errTagErr)
+	}
+
+	if clusterTagErr != nil {
+		panic(clusterTagErr)
+	}
+
+	if nodeTagErr != nil {
+		panic(nodeTagErr)
+	}
+
+	if typeTagErr != nil {
+		panic(typeTagErr)
+	}
+
+	views := []*view.View{
+		cdsRejectView,
+		edsRejectView,
+		ldsRejectView,
+		rdsRejectView,
+		edsInstancesView,
+		rdsExpiredNonceView,
+		totalXDSRejectsView,
+		monServicesView,
+		monVServicesView,
+		xdsClientsView,
+		xdsResponseWriteTimeoutsView,
+		pushTimeoutsView,
+		pushTimeoutFailuresView,
+		pushesView,
+		pushErrorsView,
+		proxiesConvergeDelayView,
+		pushContextErrorsView,
+		totalXDSInternalErrorsView,
+		inboundUpdatesView,
+	}
+
+	if err := view.Register(views...); err != nil {
+		panic(err)
+	}
```

Comment:
This is surely better.

Reply:
refactored a bit to: `monitoring.MustRegisterViews()`.